internal/credentials/kubernetes/acr/workload_identity.go

```diff
@@ -0,0 +1,330 @@
+package acr
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/confidential"
+	"github.com/patrickmn/go-cache"
+	v1 "k8s.io/api/authentication/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/credentials"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+const (
+	accessTokenUsername  = "00000000-0000-0000-0000-000000000000"
+	metadataURL          = "http://169.254.169.254/metadata/instance?api-version=2021-02-01"
+	viewerRoleName       = "kargo-viewer"
+	AzureDefaultAudience = "api://AzureADTokenExchange"
+	AnnotationClientID   = "azure.workload.identity/client-id"
+	AnnotationTenantID   = "azure.workload.identity/tenant-id"
+)
+
+type workloadIdentityCredentialHelper struct {
+	tokenCache *cache.Cache
+
+	tenantID    string
+	clientID    string
+	kargoClient client.Client
+	azureEnv    *azure.Environment
+
+	// The following behaviors are overridable for testing purposes:
+	getProjectFn    func(ctx context.Context, project string) (*v1alpha1.Project, error)
+	exchangeTokenFn func(ctx context.Context, token, clientID, oAuthEndpoint, scope string) (string, error)
+	fetchSAFn       func(ctx context.Context, project string, audiences []string) (string, error)
+	fetchACRTokenFn func(endpoint, tokenType string, data url.Values) (string, error)
+}
+
+// NewWorkloadIdentityCredentialHelper returns an implementation
+// credentials.Helper that utilizes a cache to avoid unnecessary calls to Azure.
+func NewWorkloadIdentityCredentialHelper(ctx context.Context, kargoClient client.Client) credentials.Helper {
+	logger := logging.LoggerFromContext(ctx)
+	tenantID := os.Getenv("AZURE_TENANT_ID")
+	clientID := os.Getenv("AZURE_CLIENT_ID")
+
+	env, err := getAzureEnvironment()
+	if err != nil {
+		logger.Info("Azure environment not set; Azure Workload Identity integration is disabled")
+		return nil
+	}
+	logger.Info("Azure Workload Identity integration is enabled")
+
+	w := &workloadIdentityCredentialHelper{
+		tenantID:    tenantID,
+		clientID:    clientID,
+		kargoClient: kargoClient,
+		azureEnv:    &env,
+		tokenCache: cache.New(
+			// Access tokens live for three hours. We'll hang on to them for 40 minutes.
+			40*time.Minute, // Default ttl for each entry
+			time.Hour,      // Cleanup interval
```

Comment:
If the tokens live for three hours, I would hang on to them for longer than 40 mins. 2.5 hours seems decent?